Prebuilt Binaries for Android Termux (aarch64)

## Prebuilt Binaries for Android Termux (aarch64)

Currently, prebuilt binaries for `blink.cmp` are not available for `Android Termux on aarch64` architecture. This forces users on this platform to build from source, which requires the nightly Rust toolchain. 

**Issues with Building from Source:**

Installing the nightly Rust toolchain on Termux can be challenging and error-prone.  Users frequently encounter issues related to dependencies and build tools.

**Feature Request:** Provide prebuilt binaries for Android Termux (aarch64) to eliminate the need for a nightly Rust toolchain and simplify installation and usage for users on this platform. This will significantly improve accessibility for Termux users and expand the plugin's user base.

```bash 
$ ldd target/release/libblink_cmp_fuzzy.so

        libdl.so => /system/lib64/libdl.so                                                libc.so => /system/lib64/libc.so
        ld-android.so => /system/lib64/ld-android.so
```

```bash
$ ld target/release/libblink_cmp_fuzzy.so


ld: warning: cannot find entry symbol _start; not setting start address
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_isstring'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_setmetatable
'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushfstring'ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushinteger'ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_gettable'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushvalue'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_xmove'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_createtable'ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_touserdata'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushcclosure
'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_insert'     ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushthread'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_tothread'   ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_gettop'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_typename'   ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_newuserdata'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushlightuse
rdata'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_getstack'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_rawseti'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushboolean'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_settable'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_settop'     ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `luaL_getmetafiel
d'                                                                                
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_rawget'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pcall'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_checkstack'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_topointer'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_concat'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_rawequal'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_toboolean'  ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_newthread'  ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_next'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_getinfo'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_getallocf'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_error'      ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_type'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_replace'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_isnumber'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_rawset'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `luaL_callmeta'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_remove'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushlstring'ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_rawgeti'    ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_tolstring'  ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `luaL_error'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_tonumber'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushnumber'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_getmetatable'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `lua_pushnil'
ld: target/release/libblink_cmp_fuzzy.so: undefined reference to `luaL_ref'
```

--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -15,3 +15,13 @@ rustflags = ["-C", "target-feature=-crt-static"]
 
 [target.aarch64-unknown-linux-musl]
 rustflags = ["-C", "target-feature=-crt-static"]
+
+[target.aarch64-linux-android]
+rustflags = [
+  "-C",
+  "linker=aarch64-linux-android-clang",
+  "-C",
+  "link-args=-rdynamic",
+  "-C",
+  "default-linker-libraries",
+]

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -29,6 +29,10 @@ jobs:
           - os: ubuntu-latest
             target: aarch64-unknown-linux-musl
             artifact_name: target/aarch64-unknown-linux-musl/release/libblink_cmp_fuzzy.so
+          # Android(Termux)
+          - os: ubuntu-latest
+            target: aarch64-linux-android
+            artifact_name: target/aarch64-linux-android/release/libblink_cmp_fuzzy.so
 
           ## macOS builds
           - os: macos-latest

--- a/lua/blink/cmp/fuzzy/download.lua
+++ b/lua/blink/cmp/fuzzy/download.lua
@@ -2,6 +2,21 @@ local download_config = require('blink.cmp.config').fuzzy.prebuilt_binaries
 
 local download = {}
 
+download.system_triples = {
+  mac = {
+    arm = 'aarch64-apple-darwin',
+    x64 = 'x86_64-apple-darwin',
+  },
+  windows = {
+    x64 = 'x86_64-pc-windows-msvc',
+  },
+  linux = {
+    android = 'aarch64-linux-android',
+    arm = function(libc) return 'aarch64-unknown-linux-' .. libc end,
+    x64 = function(libc) return 'x86_64-unknown-linux-' .. libc end,
+  },
+}
+
 --- @return string
 function download.get_lib_extension()
   if jit.os:lower() == 'mac' or jit.os:lower() == 'osx' then return '.dylib' end
@@ -134,41 +149,49 @@ function download.set_downloaded_version(version, cb)
   end)
 end
 
---- @param cb fun(triple: string | nil)
+--- @return string, string
+function download.get_system_info()
+  local os = jit.os:lower()
+  if os == 'osx' then os = 'mac' end
+  local arch = jit.arch:lower():match('arm') and 'arm' or jit.arch:lower():match('x64') and 'x64' or nil
+  return os, arch
+end
+
+--- @param cb fun(triple: string | function | nil)
 function download.get_system_triple(cb)
   if download_config.force_system_triple then return cb(download_config.force_system_triple) end
 
-  if jit.os:lower() == 'mac' or jit.os:lower() == 'osx' then
-    if jit.arch:lower():match('arm') then return cb('aarch64-apple-darwin') end
-    if jit.arch:lower():match('x64') then return cb('x86_64-apple-darwin') end
-  elseif jit.os:lower() == 'windows' then
-    if jit.arch:lower():match('x64') then return cb('x86_64-pc-windows-msvc') end
-  elseif jit.os:lower() == 'linux' then
+  local os, arch = download.get_system_info()
+  local triples = download.system_triples[os]
+
+  if os == 'linux' then
+    if vim.fn.has('android') == 1 then return cb(triples.android) end
+
     vim.uv.fs_stat('/etc/alpine-release', function(err, is_alpine)
       local libc = (not err and is_alpine) and 'musl' or 'gnu'
-      if jit.arch:lower():match('arm') then return cb('aarch64-unknown-linux-' .. libc) end
-      if jit.arch:lower():match('x64') then return cb('x86_64-unknown-linux-' .. libc) end
-      return cb(nil)
+      local triple = triples[arch]
+      return cb(triple and type(triple) == 'function' and triple(libc) or triple)
     end)
   else
-    return cb(nil)
+    return cb(triples[arch])
   end
 end
 
 function download.get_system_triple_sync()
   if download_config.force_system_triple then return download_config.force_system_triple end
 
-  if jit.os:lower() == 'mac' or jit.os:lower() == 'osx' then
-    if jit.arch:lower():match('arm') then return 'aarch64-apple-darwin' end
-    if jit.arch:lower():match('x64') then return 'x86_64-apple-darwin' end
-  elseif jit.os:lower() == 'windows' then
-    if jit.arch:lower():match('x64') then return 'x86_64-pc-windows-msvc' end
-  elseif jit.os:lower() == 'linux' then
+  local os, arch = download.get_system_info()
+  local triples = download.system_triples[os]
+
+  if os == 'linux' then
+    if vim.fn.has('android') == 1 then return triples.android end
+
     local success, is_alpine = pcall(vim.uv.fs_stat, '/etc/alpine-release')
     local libc = (success and is_alpine) and 'musl' or 'gnu'
-
-    if jit.arch:lower():match('arm') then return 'aarch64-unknown-linux-' .. libc end
-    if jit.arch:lower():match('x64') then return 'x86_64-unknown-linux-' .. libc end
+    local triple = triples[arch]
+    return triple and type(triple) == 'function' and triple(libc) or triple
+  else
+    return triples[arch]
   end
 end
 

